test/request.js

```diff
@@ -405,6 +405,69 @@ describe('swagger basePath override functions', function () {
     });
   });
 
+  it('applies a request interceptor per #601', function(done) {
+    var petApi = sample.pet;
+    var requestInterceptor = {
+      apply: function(data) {
+        // rewrites an invalid pet id (-100) to be valid (1)
+        // you can do what you want here, like inject headers, etc.
+        data.url = 'http://petstore.swagger.io/v2/pet/1';
+        return data;
+      }
+    };
+
+    petApi.getPetById({petId: -100}, {requestInterceptor: requestInterceptor}, function(success){
+      expect(success.obj).toBeAn('object');
+      expect(success.obj.id).toBe(1);
+      done();
+    }, function(err){
+      expect(err.obj).toBeAn('object');
+      expect(err.obj.code).toBe(400);
+      expect(err.obj.type).toBe('bad input');
+      expect(err.obj.message).toBe('sorry!');
+      done();
+    });
+  });
+
+  it('applies both a request and response interceptor per #601', function(done) {
+    var petApi = sample.pet;
+    var startTime = 0;
+    var elapsed = 0;
+
+    var interceptor = {
+      requestInterceptor: {
+        apply: function (data) {
+          // rewrites an invalid pet id (-100) to be valid (1)
+          // you can do what you want here, like inject headers, etc.
+          startTime = new Date().getTime();
+          return data;
+        }
+      },
+      responseInterceptor: {
+        apply: function(data) {
+          elapsed = new Date().getTime() - startTime;
+          return data;
+        }
+      }
+    }
+
+    petApi.getPetById({petId: 1}, {
+      requestInterceptor: interceptor.requestInterceptor,
+      responseInterceptor: interceptor.responseInterceptor
+    }, function(success){
+      expect(success.obj).toBeAn('object');
+      expect(success.obj.id).toBe(1);
+      expect(elapsed).toBeGreaterThan(0);
+      done();
+    }, function(err){
+      expect(err.obj).toBeAn('object');
+      expect(err.obj.code).toBe(400);
+      expect(err.obj.type).toBe('bad input');
+      expect(err.obj.message).toBe('sorry!');
+      done();
+    });
+  });
+
```

Comment:
Can you please add a supporting test case for #599 as you claim this patch solves for it??? I still cannot see how the API has access to the lower-level `req`, `res`, and `err` objects.